### PR TITLE
Add support for multi cluster setup in helm

### PR DESCRIPTION
**Pull request by evalsocket**
- Added support for multi cluster setup of flyte through helm
- Added docs for multi cluster setup with helm

Fix: https://github.com/flyteorg/flyte/issues/2107

Docs Deployment: https://flyte--2169.org.readthedocs.build/en/2169/deployment/multicluster.html

### CODE REVIEW

File: charts/flyte-core/values.yaml [hunk 1]
```diff
@@ -356,6 +356,16 @@ db:
...
 configmap:
+  # Details of cluster
+  clusters: []
+#    - name: "cluster_1"
+#      endpoint: { { your-cluster-1-kubeapi-endpoint.com } }
+#      enabled: true
+#      auth:
+#        type: "file_path"
+#        tokenPath: "/var/run/credentials/cluster_1_token"
+#        certPath: "/var/run/credentials/cluster_1_cacert"
```
Comment (EngHabu) [inline charts/flyte-core/values.yaml:367]: Should we add a way to add these in `values.yaml`? or is the expectation to use `secrets:` section?

Comment (evalsocket) [reply charts/flyte-core/values.yaml:367]: we don't need that, User can create secret separately before control plane deployment  and use additional Volume to mount secret in flyteadmin pod.

**Review by samhita-alla — DISMISSED** [review]

File: rsts/deployment/multicluster.rst [hunk 2]
```diff
@@ -9,16 +9,18 @@ Scaling Beyond Kubernetes
...
+  As described in the `Architecture Overview <https://docs.flyte.org/en/latest/concepts/architecture.html>`_, the Flyte ``Control Plane`` sends workflows off to the ``Data Plane`` for execution. The data plane fulfills these workflows by launching pods in Kubernetes.
 
-Often, the total compute needs could exceed the limits of a single Kubernetes cluster. 
-To address this, you can deploy the data-plane to several isolated Kubernetes clusters.
-The control-plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data-planes, which protects you from failure in a single Kubernetes cluster, increasing scalability.
+Often, the total compute needs could exceed the limits of a single Kubernetes cluster.
+To address this, you can deploy the data plane to several isolated Kubernetes clusters.
+The control-plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data planes, which protects you from failure in a single Kubernetes cluster, increasing scalability.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:16]: Is this control plane or control-plane?

Comment (AlekhyaSasi) [reply rsts/deployment/multicluster.rst:16]:
```suggestion
The control plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data planes, protecting you from failure in a single Kubernetes cluster increasing scalability.
```

File: rsts/deployment/multicluster.rst [hunk 3]
```diff
@@ -9,16 +9,18 @@ Scaling Beyond Kubernetes
...
 
-To achieve this, first, you have to create additional Kubernetes clusters. 
-For now, let’s assume you have three Kubernetes clusters and that you can access them all with ``kubectl``. Let’s call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:19]: Single quotes are curly here and in the next line they are straight.

File: rsts/deployment/multicluster.rst [hunk 4]
```diff
@@ -9,16 +9,18 @@ Scaling Beyond Kubernetes
...
+To achieve this, first, you have to create additional Kubernetes clusters.
+For now, let's assume you have three Kubernetes clusters and that you can access them all with ``kubectl``.
+Lets call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:20]:
```Suggestion
Let's call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.
```

File: rsts/deployment/multicluster.rst [hunk 5]
```diff
@@ -52,10 +54,18 @@ Data Plane Deployment
 User and Control Plane Deployment
 *********************************
 
-Some deployments of Flyte may choose to run the control-plane separate from the data-plane. FlyteAdmin is designed to create Kubernetes resources in one or more Flyte data-plane clusters. For the admin to access remote clusters, it needs credentials to each cluster.
-In Kubernetes, scoped service credentials are created by configuring a “Role” resource in a Kubernetes cluster. When you attach that role to a “ServiceAccount”, Kubernetes generates a bearer token that permits access. We create a FlyteAdmin `ServiceAccount <https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/templates/admin/rbac.yaml#L4>`_ in each data-plane cluster to generate these tokens.
+Some deployments of Flyte may choose to run control plane separate from the data plane.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:57]:
```suggestion
Some Flyte deployments may choose to run the control plane separate from the data plane.
```

File: rsts/deployment/multicluster.rst [hunk 6]
```diff
@@ -52,10 +54,18 @@ Data Plane Deployment
...
+FlyteAdmin is designed to create Kubernetes resources in one or more Flyte data plane clusters.
+For the admin to access remote clusters, it needs credentials to each cluster.
 
-When you first create the FlyteAdmin ServiceAccount in a new cluster, a bearer token is generated and will continue to allow access unless the “ServiceAccount“ is deleted. Once we create the FlyteAdmin ServiceAccount on a cluster, we should never delete it. To feed the credentials to FlyteAdmin, you must retrieve them from your new data-plane cluster, and upload them to admin somehow (within Lyft, we use Confidant, for example).
+In Kubernetes, scoped service credentials are created by configuring a “Role” resource in a Kubernetes cluster.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:61]: Please check the curly vs. straight quotes consistency.

File: rsts/deployment/multicluster.rst [hunk 7]
```diff
@@ -52,10 +54,18 @@ Data Plane Deployment
...
+When you attach the role to a “ServiceAccount”, Kubernetes generates a bearer token that permits access.
+Hence, create a FlyteAdmin `ServiceAccount <https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/templates/admin/rbac.yaml#L4>`_ in each data plane cluster to generate these tokens.
+
+When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the “ServiceAccount“ is deleted.
+Hence, you should never delete a ``ServiceAccount`` ⚠️.
+
+To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster, and upload them to admin somehow (for example, within Lyft, :ref:`Confidant <https://github.com/lyft/confidant>`__ is used).
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:68]: This sounds informal. Could we remove somehow?

Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:68]:
```suggestion
To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster and upload them to admin (for example, within Lyft, ref:`Confidant <https://github.com/lyft/confidant>`__ is used).
```

File: rsts/deployment/multicluster.rst (hunk 2 above)
Comment (samhita-alla) [reply rsts/deployment/multicluster.rst:16]: I think it has to be "control plane". Let's use this for now.

Also, let's make the following sentence clear:

... protecting you from failure in a single Kubernetes cluster, thus increasing scalability.

**Comment by samhita-alla** [comment]
@AlekhyaSasi, would you mind pushing your suggestions to this PR?

File: rsts/deployment/multicluster.rst [hunk 8]
```diff
@@ -9,59 +9,79 @@ Scaling Beyond Kubernetes
...
 *********************************
 
-For FlyteAdmin to create Flyte Workflows on the three remote clusters, it will need a secret ``token`` and ``cacert`` to access each cluster.
+Some deployments of Flyte may choose to run control plane separate from the data plane.
+FlyteAdmin is designed to create Kubernetes resources in one or more Flyte data plane clusters.
+For the admin to access remote clusters, it needs credentials to each cluster.
 
-Once you have deployed the data-plane as described above, you can retrieve the ``token`` and ``cacert`` by pointing your ``kubectl`` context to each data-plane cluster and executing the following commands:
+In Kubernetes, scoped service credentials are created by configuring a “Role” resource in a Kubernetes cluster.
+When you attach the role to a “ServiceAccount”, Kubernetes generates a bearer token that permits access.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:62]:
```suggestion
When you attach the role to a "ServiceAccount", Kubernetes generates a bearer token that permits access.
```

File: rsts/deployment/multicluster.rst [hunk 9]
```diff
@@ -9,59 +9,79 @@ Scaling Beyond Kubernetes
...
+Hence, create a FlyteAdmin `ServiceAccount <https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/templates/admin/rbac.yaml#L4>`_ in each data plane cluster to generate these tokens.
 
-:token:
-  ``kubectl get secrets -n flyte | grep flyteadmin-token | awk '{print $1}' | xargs kubectl get secret -n flyte -o jsonpath='{.data.token}'``
+When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the “ServiceAccount“ is deleted.
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:65]:
```suggestion
When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the "ServiceAccount" is deleted.
```

File: rsts/deployment/multicluster.rst [hunk 10]
```diff
@@ -9,59 +9,79 @@ Scaling Beyond Kubernetes
...
+Hence, you should never delete a ``ServiceAccount`` ⚠️.
 
-:cacert:
-  ``kubectl get secrets -n flyte | grep flyteadmin-token | awk '{print $1}' | xargs kubectl get secret -n flyte -o jsonpath='{.data.ca\.crt}'``
+To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster, and upload them to admin somehow (for example, within Lyft, :ref:`Confidant <https://github.com/lyft/confidant>`__ is used).
 
-Now, these credentials need to be included in the control-plane. Create a new file named ``admindeployment/secrets.yaml`` that looks like: ::
+The credentials have two parts (“ca cert“ and “bearer token"). Find the generated secret via, ::
```
Comment (AlekhyaSasi) [inline rsts/deployment/multicluster.rst:70]:
```suggestion
The credentials have two parts ("ca cert" and "bearer token"). Find the generated secret via,::
```

**Review by AlekhyaSasi — APPROVED** [review]

File: rsts/deployment/multicluster.rst [hunk 11]
```diff
@@ -9,148 +9,192 @@ Scaling Beyond Kubernetes
...
+.. tabbed:: GCP
 
-  kubectl get secrets -n flyte | grep flyteadmin-token
+    .. code-block::
 
-Once you have the name of the secret, you can copy the ``ca cert`` to your clipboard using the following command: ::
+        helm upgrade flyte -n flyte flyteorg/flyte-core values.yaml -f values-gcp.yaml -f values-controlplane.yaml -f values-override.yaml --create-namespace flyte --install
 
-  kubectl get secret -n flyte {secret-name} -o jsonpath='{.data.ca\.crt}' | base64 -D | pbcopy
+Configure Execution Cluster Labels
```
[the remark this answers is not in the document]
Comment (samhita-alla) [reply rsts/deployment/multicluster.rst:180]: Added the content.

File: rsts/deployment/multicluster.rst [hunk 12]
```diff
@@ -9,148 +9,192 @@ Scaling Beyond Kubernetes
...
+        labelClusterMap:
+          team1:
+            - id: cluster_1
+              weight: 1
+          team2:
+            - id: cluster_2
+              weight: 0.5
+            - id: cluster_3
+              weight: 0.5
+        clusterConfigs:
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply rsts/deployment/multicluster.rst]: Fix the values

File: deployment/eks/flyte_helm_controlplane_generated.yaml [hunk 13]
```diff
@@ -0,0 +1,1249 @@
...
+    - http:
+        paths:
+          - path: /*
+            pathType: ImplementationSpecific
+            backend:
+              service:
+                name: ssl-redirect
+                port:
+                  name: use-annotation
+          # This is useful only for frontend development
```
Comment (AlekhyaSasi) [inline deployment/eks/flyte_helm_controlplane_generated.yaml:1008]:
```suggestion
          # This is useful only for frontend development.
```

File: deployment/eks/flyte_helm_controlplane_generated.yaml [hunk 14]
```diff
@@ -0,0 +1,1249 @@
...
+          # Port 87 in FlyteAdmin maps to the redoc container.
+          - path: /openapi
+            pathType: ImplementationSpecific
+            backend:
+              service:
+                name: flyteadmin
+                port:
+                  number: 87
+          # NOTE: If you change this, you must update the BASE_URL value in flyteconsole.yaml
```
Comment (AlekhyaSasi) [inline deployment/eks/flyte_helm_controlplane_generated.yaml:1017]:
```suggestion
          # NOTE: If you change this, you must update the BASE_URL value in flyteconsole.yaml.
```

File: deployment/eks/flyte_helm_controlplane_generated.yaml [hunk 15]
```diff
@@ -0,0 +1,1249 @@
...
+          - path: /oauth2/*
+            pathType: ImplementationSpecific
+            backend:
+              service:
+                name: flyteadmin
+                port:
+                  number: 80
+      host: null
+# Certain ingress controllers like nginx cannot serve HTTP 1 and GRPC with a single ingress because GRPC can only
+# enabled on the ingress object, not on backend services (GRPC annotation is set on the ingress, not on the services).
```
Comment (AlekhyaSasi) [inline deployment/eks/flyte_helm_controlplane_generated.yaml:1161]:
```suggestion
# be enabled on the ingress object, not on backend services (GRPC annotation is set on the ingress, not on the services).
```

Comment (evalsocket) [reply deployment/eks/flyte_helm_controlplane_generated.yaml]: @AlekhyaSasi these are generated file from helm, If you want to change them then please update these comment in values file of [flyte](https://github.com/flyteorg/flyte/blob/master/charts/flyte/values.yaml) & [flyte-core](https://github.com/flyteorg/flyte/tree/master/charts/flyte-core)

Comment (evalsocket) [reply deployment/eks/flyte_helm_controlplane_generated.yaml]: and then run `make helm`

File: charts/flyte-core/values-dataplane.yaml [hunk 16]
```diff
@@ -0,0 +1,17 @@
+flyteadmin:
+  enabled: false
+
+cluster_resource_manager:
```
Comment (EngHabu) [inline charts/flyte-core/values-dataplane.yaml:4]: This should move to the controlplane.yaml

**Review by EngHabu — COMMENTED** [review]
I've one comment... then we can merge!

File: charts/flyte-core/values-dataplane.yaml (hunk 16 above)
Comment (evalsocket) [reply charts/flyte-core/values-dataplane.yaml:4]: it will just disable the cluster_resource_manager in dataplane

**Review by EngHabu — APPROVED** [review]

**Pull request merged**